File: app/src/main/java/com/example/spelltracker/SpellRepository.java
package com.example.spelltracker;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Единая точка доступа к данным заклинаний.
 *
 * <p>Логика: при первом обращении таблица пуста — данные импортируются из
 * {@code assets/spells.csv} (с транзакцией, чтобы ~480 вставок были быстрыми).
 * Все дальнейшие запросы идут через {@link SpellDao}.</p>
 *
 * <p>Все методы асинхронные, колбэки вызываются на UI-потоке.</p>
 */
public class SpellRepository {

    /** Простой колбэк. */
    public interface Callback<T> {
        void onResult(T result);
    }

    private final SpellDao dao;
    private final SpellDatabase db;
    private final Context appContext;
    private final ExecutorService io = Executors.newSingleThreadExecutor();
    private final Handler main = new Handler(Looper.getMainLooper());

    public SpellRepository(Context context) {
        this.appContext = context.getApplicationContext();
        this.db = SpellDatabase.get(appContext);
        this.dao = db.spellDao();
    }

    /**
     * Инициализирует БД: если таблица пуста — импортирует из CSV.
     * Колбэк получает общее число записей после инициализации.
     */
    public void loadOrImport(Callback<Integer> onReady) {
        io.execute(() -> {
            try {
                if (dao.count() == 0) {
                    importFromCsv();
                }
            } catch (Throwable t) {
                t.printStackTrace();
            }
            final int c = dao.count();
            main.post(() -> {
                if (onReady != null) onReady.onResult(c);
            });
        });
    }

    public void getAll(Callback<List<Spell>> cb) {
        io.execute(() -> {
            final List<Spell> result = dao.getAll();
            main.post(() -> {
                if (cb != null) cb.onResult(result);
            });
        });
    }

    public void getByLevel(int level, Callback<List<Spell>> cb) {
        io.execute(() -> {
            final List<Spell> result = dao.getByLevel(level);
            main.post(() -> {
                if (cb != null) cb.onResult(result);
            });
        });
    }

    public void getById(long id, Callback<Spell> cb) {
        io.execute(() -> {
            final Spell s = dao.getById(id);
            main.post(() -> {
                if (cb != null) cb.onResult(s);
            });
        });
    }

    /** Синхронный импорт CSV в БД (вызывается в фоновом потоке). */
    private void importFromCsv() throws IOException {
        List<Spell> all = SpellParser.loadFromAssets(appContext);
        db.runInTransaction(() -> {
            for (Spell s : all) {
                dao.insert(s);
            }
        });
    }
}

File: app/src/main/java/com/example/spelltracker/data/SpellRepository.kt
package com.example.spelltracker.data

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Единая точка доступа к Room-базе со справочником заклинаний.
 *
 * Приложение запускает [ensureInitialized] при старте, и если БД пуста —
 * парсит `spells_normalized.json` из assets и заливает все заклинания в таблицу
 * `spells`. Данные собраны build-time Gradle-таской `generateSpellsDb` из
 * per-spell JSON в `spells_data/` (см. app/build.gradle.kts).
 * `spells`. После этого [initialized] становится true, и ViewModel-ы
 * могут реагировать.
 */
class SpellRepository(context: Context) {

    private val appContext = context.applicationContext
    private val db = SpellDatabase.get(appContext)
    private val dao = db.spellDao()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _initialized = MutableStateFlow(false)
    val initialized: StateFlow<Boolean> = _initialized.asStateFlow()

    fun ensureInitialized() {
        scope.launch {
            if (dao.count() == 0) {
                val allSpells = SpellParser.loadFromAssets(appContext)
                if (allSpells.isNotEmpty()) {
                    dao.insertAll(allSpells)
                }
            }
            _initialized.value = true
        }
    }

    suspend fun getAll(): List<Spell> = withContext(Dispatchers.IO) { dao.getAll() }

    suspend fun getById(id: Long): Spell? = withContext(Dispatchers.IO) { dao.getById(id) }

    /**
     * Загрузить все заклинания и отфильтровать по UI-параметрам.
     * Фильтрация делается в памяти — справочник маленький, JSON-ы
     * кешируются Room-ом, так что это дёшево.
     */
    suspend fun filter(
        level: Int?,            // null = любой
        classIds: Set<String>,  // пусто = любой
        search: String,         // "" = без поиска по имени
    ): List<Spell> = withContext(Dispatchers.IO) {
        val all = dao.getAll()
        val needle = search.trim().lowercase()
        all.asSequence()
            .filter { level == null || it.level == level }
            .filter { classIds.isEmpty() || classIds.any { id -> it.classes.contains(id) } }
            .filter { needle.isEmpty() || it.name.lowercase().contains(needle) }
            .toList()
    }
}
